src/plugins/intel_gna/memory/gna_memory.hpp

```diff
@@ -14,63 +14,75 @@
 #include <algorithm>
 #include <functional>
 #include <iostream>
+#include <fstream>
+#include <utility>
 #include "gna_lib_ver_selector.hpp"
 #include "memory_solver.hpp"
+#include "gna_allocator.hpp"
 #include "gna_plugin_log.hpp"
 #include "memory/gna_allocator.hpp"
 
 #ifdef GNA_HEAP_PROFILER
 #include <iomanip>
-#include <fstream>
 #endif
 
 namespace GNAPluginNS {
 namespace memory {
+
+class GNAFloatAllocator : public std::allocator < uint8_t > {
+ public:
+    void setTag(void*, GNAPluginNS::memory::rRegion) {
+    }
+};
+
+class GNAMemoryInterface {
+public:
+    virtual GNAMemRequestsQueue* getQueue(rRegion region) = 0;
+    virtual GNAMemRequestsQueue* getQueue(void* ptr) = 0;
+    virtual void commit(bool isCompact = false) = 0;
+    virtual std::pair<bool, uint32_t> getOffsetForMerged(void* ptr) = 0;
+    virtual size_t getRegionBytes(rRegion region) = 0;
+    virtual ~GNAMemoryInterface() = default;
+};
+
 /**
  * @brief encapsulate various request to allocate GNA specific memory,
  * in order to issue single allocation call and configure actual pointers in requests
  * @tparam Allocator - a GNAAllocator in case of actual HW offloads
  */
-template<class Allocator = std::allocator<uint8_t>>
-class GNAMemory : public GNAMemRequestsQueue {
+template<class Allocator = GNAAllocator>
+class GNAMemory : public GNAMemoryInterface {
 protected:
-    std::vector<MemRequest> _future_heap;
-    std::list<std::vector<char>> _local_storage;
+    std::map<rRegion, std::unique_ptr<GNAMemRequestsQueue>> _mem_queues;
     size_t _total = 0;
-    size_t _rw_section_size = 0;
-    size_t _ro_section_size = 0;
     Allocator _allocator;
-    std::shared_ptr<uint8_t> heap = nullptr;
     size_t _page_alignment = 1;
     bool _is_compact_mode = false;
 
-    class GNAMemRequestsReadOnlyQueue : public GNAMemRequestsQueue {
-        std::reference_wrapper<GNAMemRequestsQueue> _that;
-     public:
-        explicit GNAMemRequestsReadOnlyQueue(GNAMemory & that) : _that(that) {
-        }
-        rRegion regionType() const override {
-            return REGION_RO;
-        };
-        std::vector<MemRequest> & futureHeap()  override {
-            return _that.get().futureHeap();
-        }
-        std::list<std::vector<char>> &localStorage() override {
-            return _that.get().localStorage();
-        }
-    };
-
-    GNAMemRequestsReadOnlyQueue readOnlyFrontEnd;
+ private:
+    void initMemQueses() {
+        _mem_queues.insert(std::make_pair(REGION_RO, new GNAMemRequestsReadOnlyQueue()));
```

Comment:
minor: why not use `make_unique` wrapper here?

Reply:
I can change it in the next PR if you don't mind (btw, AFAIK current OV cpp standard does not have `make_unique` available anyway so I would have to reuse some custom implementation).
@mbencer is it OK?

Reply:
@kbruniec  that's right, as I remember it was introduced in cpp14, but I mean about wrapper which you added in https://github.com/openvinotoolkit/openvino/blob/1d33a5c5d1ad7a90da5acfcc9063f10181ac28e1/src/plugins/intel_gna/backend/gna_limitations.cpp#L22. Nevertheless, I am not familiar with with GNA dependency, so maybe it is a bad idea.
Anyway, it's a minor comment. If @t-jankowski review it, we can merge ;)

Reply:
@mbencer Reviews were applied (including make_unique issue) and approved. Ready to be merged.